subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/excludes/factories/NormalizingExcludeFactory.java

```diff
@@ -0,0 +1,247 @@
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.excludes.specs.GroupExclude;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.excludes.specs.GroupSetExclude;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.excludes.specs.ModuleExclude;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.excludes.specs.ModuleIdExclude;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.excludes.specs.ModuleIdSetExclude;
+import org.gradle.api.internal.artifacts.ivyservice.resolveengine.excludes.specs.ModuleSetExclude;
+import org.gradle.internal.Cast;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * This factory performs normalization of exclude rules. This is the smartest
+ * of all factories and is responsible for doing some basic algebra computations.
+ * It shouldn't be too slow, or the whole chain will pay the price.
+ */
+public class NormalizingExcludeFactory extends DelegatingExcludeFactory {
+    public NormalizingExcludeFactory(ExcludeFactory delegate) {
+        super(delegate);
+    }
+
+    @Override
+    public ExcludeSpec anyOf(ExcludeSpec one, ExcludeSpec two) {
+        return doUnion(ImmutableSet.of(one, two));
+    }
+
+    @Override
+    public ExcludeSpec allOf(ExcludeSpec one, ExcludeSpec two) {
+        return doIntersect(ImmutableSet.of(one, two));
+    }
+
+    @Override
+    public ExcludeSpec anyOf(Set<ExcludeSpec> specs) {
+        return doUnion(specs);
+    }
+
+    @Override
+    public ExcludeSpec allOf(Set<ExcludeSpec> specs) {
+        return doIntersect(specs);
+    }
+
+    private ExcludeSpec doUnion(Set<ExcludeSpec> specs) {
+        FlattenOperationResult flattened = flatten(ExcludeAnyOf.class, specs, ExcludeEverything.class::isInstance, ExcludeNothing.class::isInstance);
+        if (flattened.fastExit) {
+            return everything();
+        }
+        if (flattened.result.isEmpty()) {
+            return nothing();
+        }
+        Map<UnionOf, List<ExcludeSpec>> byType = flattened.result.stream().collect(Collectors.groupingBy(UnionOf::typeOf));
+        List<ModuleIdExclude> moduleIdExcludes = UnionOf.MODULEID.fromMap(byType);
+        List<ModuleIdSetExclude> moduleIdSetsExcludes = UnionOf.MODULEID_SET.fromMap(byType);
+        List<GroupExclude> groupExcludes = UnionOf.GROUP.fromMap(byType);
+        List<GroupSetExclude> groupSetExcludes = UnionOf.GROUP_SET.fromMap(byType);
+        List<ModuleExclude> moduleExcludes = UnionOf.MODULE.fromMap(byType);
+        List<ModuleSetExclude> moduleSetExcludes = UnionOf.MODULE_SET.fromMap(byType);
+        List<ExcludeSpec> other = UnionOf.NOT_JOINABLE.fromMap(byType);
+        if (!moduleIdExcludes.isEmpty()) {
+            if (moduleIdExcludes.size() > 1 || !moduleIdSetsExcludes.isEmpty()) {
```

Comment:
A small comment to express intent might help there, or extract to a method.
> // More than one ModuleIdExclude -> make it a ModuleIdSetExclude

Also, with the flattening done below, there might be a way to optimize when the `moduleIdSetsExcludes.isEmpty()` is true as you now create an immutable list that will be flattened again.

Reply:
This code path is not called often and doesn't show up as a hotspot, so I didn't bother optimizing yet.